docs/development/architecture.org

```diff
@@ -0,0 +1,130 @@
+Nix-Gui may be considered to have two components: the backend and the frontend.
+
+The backend (=nixui/options=) reads information from both NixOS and the configuration path being edited, then converts them to data-structures useful for navigation, search, and editing. The backend maintains in-memory state for all changes and commits those changes to disk by writing to the configuration path.
```

Comment:
> reads information from both NixOS and the configuration path being edited

reads information from two sources:

1. from NixOS (where exactly?)
2. from the configuration path being edited

> then converts

is this conversion lazy?

> The backend maintains in-memory state for all changes and commits those changes to disk by writing to the configuration path.

can we use "micro commits" to track ALL changes with git? = auto save

Reply:
All NixOS options are loaded and `pkgs.nixosOptionsDoc` is called for each option when it starts up. However `OptionDefinition.obj` is calculated at runtime. So I suppose it is lazy in the sense that the `OptionTree` is loaded when the program starts, but the Python representation of some metadata is calculated on the fly?

I think that qualifies as lazy, so I put it in for now.

First two changes resolved by https://github.com/nix-gui/nix-gui/pull/214/commits/411745c43431c5a64977e8b5eb794f2141975921

>can we use "micro commits" to track ALL changes with git? = auto save

I think a simple stack of `Update`s is the best option for linear Undo capabilities.

For committed changes, we should use git, which is part of milestone 3, but has no issue yet https://github.com/nix-gui/nix-gui/milestone/3

Reply:
> All NixOS options are loaded

so its not lazy

the options schema is finite and acyclic, so no need for lazy eval

Reply:
I'm not sure what you mean, I was thinking about lazy dynamic programming solutions which exhaust the entire search space to find an answer, but you must mean something else.

I'll remove "lazily".

Reply:
lazy as in "lazy functional programming language" (nix, haskell, ...)

aka "eval on demand"

here we need a full eval, to build a search index for all paths of he config schema